File: backend/src/main/java/bingol/campus/comment/business/abstracts/CommentService.java
package bingol.campus.comment.business.abstracts;

import bingol.campus.comment.core.exception.CommentNotFoundException;
import bingol.campus.comment.core.exception.UnauthorizedCommentException;
import bingol.campus.comment.core.response.CommentDTO;
import bingol.campus.followRelation.core.exceptions.BlockingBetweenStudent;
import bingol.campus.post.core.exceptions.PostNotFoundException;
import bingol.campus.post.core.exceptions.PostNotIsActiveException;
import bingol.campus.response.DataResponseMessage;
import bingol.campus.response.ResponseMessage;
import bingol.campus.story.core.exceptions.NotFollowingException;
import bingol.campus.story.core.exceptions.StoryNotActiveException;
import bingol.campus.story.core.exceptions.StoryNotFoundException;
import bingol.campus.story.core.exceptions.StudentProfilePrivateException;
import bingol.campus.student.exceptions.StudentNotFoundException;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface CommentService {
    ResponseMessage addCommentToStory(String username, UUID storyId, String content) throws StudentNotFoundException, StoryNotFoundException, StoryNotActiveException, BlockingBetweenStudent, NotFollowingException, StudentProfilePrivateException;

    ResponseMessage addCommentToPost(String username, UUID postId, String content) throws PostNotFoundException, StudentNotFoundException, PostNotIsActiveException, NotFollowingException, BlockingBetweenStudent, StudentProfilePrivateException;

    ResponseMessage deleteComment(String username, UUID commentId) throws CommentNotFoundException, StudentNotFoundException, UnauthorizedCommentException;

    DataResponseMessage<CommentDTO> getCommentDetails(String username, UUID commentId) throws UnauthorizedCommentException, CommentNotFoundException, StudentNotFoundException;


    DataResponseMessage<List<CommentDTO>> searchUserInStoryComments(String username, UUID storyId, String username1) throws UnauthorizedCommentException, StudentNotFoundException, StoryNotFoundException;

    DataResponseMessage<List<CommentDTO>> searchUserInPostComments(String username, UUID postId, String username1) throws PostNotIsActiveException, NotFollowingException, BlockingBetweenStudent, UnauthorizedCommentException, StudentNotFoundException, PostNotFoundException;

    DataResponseMessage<List<CommentDTO>> getUserComments(String username, Pageable pageRequest) throws StudentNotFoundException;

    DataResponseMessage<List<CommentDTO>> getStoryComments(String username, UUID storyId, Pageable pageable) throws NotFollowingException, BlockingBetweenStudent, StoryNotActiveException, StudentNotFoundException, StoryNotFoundException, StudentProfilePrivateException;

    DataResponseMessage<List<CommentDTO>> getPostComments(String username, UUID postId, Pageable pageable) throws StudentNotFoundException, PostNotFoundException, PostNotIsActiveException, NotFollowingException, BlockingBetweenStudent, StudentProfilePrivateException;
}

File: frontend/lib/services/commentService.dart
import 'package:dio/dio.dart';

class CommentService {
  final Dio _dio;

  CommentService(this._dio) {
    _dio.options.baseUrl =
        'http://192.168.89.61:8080/v1/api'; // Base URL ayarlandı
  }

  // Hikayeye yorum yapma
  Future<Response> addCommentToStory(
      String token, String storyId, String content) async {
    final response = await _dio.post(
      '/comments/story/$storyId',
      data: {'content': content},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Gönderiye yorum yapma
  Future<Response> addCommentToPost(
      String token, String postId, String content) async {
    try {
      if (content.isEmpty) {
        throw Exception("Comment content cannot be empty");
      }

      final response = await _dio.post(
        '/comments/post/$postId?content=${Uri.encodeComponent(content)}',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      if (response.statusCode == 200) {
        return response;
      } else {
        throw Exception(
            "Failed to add comment: ${response.statusCode} - ${response.data}");
      }
    } catch (e) {
      print('Error adding comment: $e');
      rethrow;
    }
  }

  // Yorum silme
  Future<Response> deleteComment(String token, String commentId) async {
    final response = await _dio.delete(
      '/comments/$commentId',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Kullanıcının yaptığı yorumları sayfalı olarak listeleme
  Future<Response> getUserComments(String token, int page) async {
    final response = await _dio.get(
      '/comments/user',
      queryParameters: {'page': page},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Belirli bir hikayedeki yorumları sayfalı olarak listeleme
  Future<Response> getStoryComments(
      String token, String storyId, int page) async {
    final response = await _dio.get(
      '/comments/story/$storyId',
      queryParameters: {'page': page},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Belirli bir gönderideki yorumları sayfalı olarak listeleme
  Future<Response> getPostComments(String token, String postId, int page,
      {int limit = 10}) async {
    final response = await _dio.get(
      '/comments/post/$postId',
      queryParameters: {
        'page': page,
        'limit': limit,
      },
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Belirli bir yorumun detaylarını getirme
  Future<Response> getCommentDetails(String token, String commentId) async {
    final response = await _dio.get(
      '/comments/$commentId',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Belli bir hikayede kullanıcı yorumu arama
  Future<Response> searchUserInStoryComments(
      String token, String storyId, String username) async {
    final response = await _dio.get(
      '/comments/story/$storyId/search/$username',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }

  // Belli bir gönderide kullanıcı yorumu arama
  Future<Response> searchUserInPostComments(
      String token, String postId, String username) async {
    final response = await _dio.get(
      '/comments/post/$postId/search/$username',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return response;
  }
}
